File: app/src/main/java/com/mmdev/meetapp/ui/auth/AuthActivity.java
package com.mmdev.meetapp.ui.auth;

import android.app.Dialog;
import android.content.Intent;
import android.graphics.Color;
import android.os.Bundle;
import android.view.Gravity;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;
import android.widget.*;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.ViewModelProviders;
import butterknife.BindView;
import butterknife.ButterKnife;
import co.ceryle.segmentedbutton.SegmentedButtonGroup;
import com.facebook.AccessToken;
import com.facebook.CallbackManager;
import com.facebook.FacebookCallback;
import com.facebook.FacebookException;
import com.facebook.login.LoginResult;
import com.facebook.login.widget.LoginButton;
import com.google.android.gms.tasks.Task;
import com.google.firebase.auth.*;
import com.google.firebase.firestore.CollectionReference;
import com.google.firebase.firestore.DocumentSnapshot;
import com.google.firebase.firestore.FirebaseFirestore;
import com.mmdev.meetapp.R;
import com.mmdev.meetapp.models.ProfileModel;
import com.mmdev.meetapp.services.GPSTracker;
import com.mmdev.meetapp.ui.main.MainActivity;
import com.mmdev.meetapp.ui.main.ProfileViewModel;
import com.mmdev.meetapp.utils.uiUtils;
import com.mmdev.progressbuttonlib.ProgressButton;

import java.util.ArrayList;


/**
 *
 */
public class AuthActivity extends AppCompatActivity
{

	private static final String TAG = "logi";

	@BindView(R.id.facebook_login_button)
	LoginButton facebookLogInButton;

	@BindView(R.id.facebook_login_button_delegate)
	Button facebookLoginButtonDelegate;

	private FirebaseAuth mFirebaseAuth;
	private FirebaseAuth.AuthStateListener mAuthStateListener;
	private FirebaseFirestore mFirestore;

	private FirebaseUser mFirebaseUser;
	//Progress dialog for any authentication action
	private AlertDialog progressDialog;

	CallbackManager mCallbackManager;

	private GPSTracker gps = new GPSTracker();
	private String mGender = "male";
	private String mPreferedGender = "male";
	private String mCity = "";
	private ProfileModel mProfileModel;

	private ProgressButton pb_button;
	private Dialog additionalRegDialog;

	private ProfileViewModel profileViewModel;

	@Override
	protected void onCreate (@Nullable Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		setContentView(R.layout.activity_auth);
		ButterKnife.bind(this);
		mFirebaseAuth = FirebaseAuth.getInstance();
		mAuthStateListener = ((@NonNull FirebaseAuth firebaseAuth) -> {});
		mFirestore = FirebaseFirestore.getInstance();
		mCallbackManager = CallbackManager.Factory.create();
		setUpFacebookLoginButton();
		setProgressDialog();
		profileViewModel = ViewModelProviders.of(this).get(ProfileViewModel.class);

	}

	private void setUpFacebookLoginButton() {
		facebookLogInButton.registerCallback(mCallbackManager, new FacebookCallback<LoginResult>() {
			@Override
			public void onSuccess (LoginResult loginResult) { handleFacebookAccessToken(loginResult.getAccessToken()); }

			@Override
			public void onCancel () {}

			@Override
			public void onError (FacebookException error) {}
		});
		facebookLoginButtonDelegate.setOnClickListener((View v)-> facebookLogInButton.performClick());
	}

	private void handleFacebookAccessToken (AccessToken token) {
		showProcessProgressDialog();
		AuthCredential credential = FacebookAuthProvider.getCredential(token.getToken());
		mFirebaseAuth.signInWithCredential(credential).addOnCompleteListener(this, ((@NonNull Task<AuthResult> task) -> {
			if (task.isSuccessful() && mFirebaseAuth.getCurrentUser() != null){
				mFirebaseUser = mFirebaseAuth.getCurrentUser();
				handleUserExistence(mFirebaseUser.getUid());
			}
			else uiUtils.showSafeToast(getApplicationContext(),"LogIn Aborted");
			dismissProgressDialog();
		}));
	}

	private void handleUserExistence (String uId){
		CollectionReference users = mFirestore.collection("users");
		users.document(uId).get().addOnCompleteListener(task -> {
			if (task.isSuccessful()) {
				DocumentSnapshot document = task.getResult();
				if (document != null && document.exists()) {
					mProfileModel = document.toObject(ProfileModel.class);
					profileViewModel.saveProfile(getApplicationContext(), mProfileModel);
					profileViewModel.setProfileModel(mProfileModel);
					Toast.makeText(getApplicationContext(), "Exist",Toast.LENGTH_SHORT).show();
					startMainActivity ();
				} else showRegistrationDialog();
			} else Toast.makeText(getApplicationContext(), "Can't connect to server",Toast.LENGTH_SHORT).show();
		});
	}

	private void showRegistrationDialog () {
		additionalRegDialog = new Dialog(this);
		additionalRegDialog.setContentView(R.layout.dialog_registration);
		Window window = additionalRegDialog.getWindow();
		additionalRegDialog.show();
		if (window != null)
			window.setLayout(LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.MATCH_PARENT);

		SegmentedButtonGroup sBG_gender = additionalRegDialog.findViewById(R.id.dialog_registr_sbg_gender);
		SegmentedButtonGroup sBG_prefGender = additionalRegDialog.findViewById(R.id.dialog_registr_sbg_preferedgender);
		
		sBG_gender.setOnClickedButtonListener(position -> {
			if (position == 0) mGender = "male";
			else mGender = "female";
		});
		
		sBG_prefGender.setOnClickedButtonListener(position -> {
			switch (position){
				case 0:
					mPreferedGender = "male";
					break;
				case 1:
					mPreferedGender = "female";
					break;
				case 2:
					mPreferedGender = "both";
					break;
			}
		});
		pb_button = additionalRegDialog.findViewById(R.id.diag_reg_btn_done);
		pb_button.setOnClickListener(v -> {
			//new AddUserToFirestore(this).execute();
            pb_button.startAnim();
            addUserToFirestore();
		});
	}

	private void addUserToFirestore () {
		//city = gps.getCity(this);
		mCity = "Kyiv";
		ArrayList<String> urls = new ArrayList<>();
		urls.add(String.valueOf(mFirebaseUser.getPhotoUrl()));
		mProfileModel = new ProfileModel(mFirebaseUser.getDisplayName(), mCity, mGender,
		                                 mPreferedGender, String.valueOf(mFirebaseUser.getPhotoUrl()),
		                                 urls, mFirebaseUser.getUid());
		CollectionReference profiles = mFirestore.collection("users");
		profiles.document(mFirebaseUser.getUid()).set(mProfileModel).addOnSuccessListener(aVoid -> {
			profileViewModel.saveProfile(this, mProfileModel);
			profileViewModel.setProfileModel(mProfileModel);
			Toast.makeText(this,
					"Successfully uploaded data to firebase + stored into sharedPrefs",
					Toast.LENGTH_LONG).show();
			pb_button.stopAnim(this::startMainActivity);
		});

	}

	private void startMainActivity () {
		Intent mMainActivityIntent = new Intent(AuthActivity.this, MainActivity.class);
		startActivity(mMainActivityIntent);
		if (additionalRegDialog!=null) additionalRegDialog.dismiss();
		finish();
	}

	/*
	progress dialog
	 */
	public void setProgressDialog() {
		int llPadding = 10;
		LinearLayout ll = new LinearLayout(this);
		ll.setOrientation(LinearLayout.HORIZONTAL);
		ll.setPadding(llPadding, llPadding, llPadding, llPadding);
		ll.setGravity(Gravity.CENTER);
		LinearLayout.LayoutParams llParam = new LinearLayout.LayoutParams(
				LinearLayout.LayoutParams.WRAP_CONTENT,
				LinearLayout.LayoutParams.WRAP_CONTENT);
		llParam.gravity = Gravity.CENTER;
		ll.setLayoutParams(llParam);
		
		ProgressBar progressBar = new ProgressBar(this);
		progressBar.setIndeterminate(true);
		progressBar.setPadding(llPadding, llPadding, llPadding, llPadding);
		progressBar.setLayoutParams(llParam);
		
		TextView tvText = new TextView(this);
		tvText.setText(getString(R.string.progress_dialog_text));
		tvText.setTextColor(Color.BLACK); //low api
		tvText.setTextSize(20);
		tvText.setLayoutParams(llParam);
		
		ll.addView(progressBar);
		ll.addView(tvText);
		
		AlertDialog.Builder builder = new AlertDialog.Builder(this);
		builder.setCancelable(true);
		builder.setView(ll);
		
		progressDialog = builder.create();
		
	}

	private void showProcessProgressDialog () {
		progressDialog.show();
		Window window = progressDialog.getWindow();
		if (window != null) {
			WindowManager.LayoutParams layoutParams = new WindowManager.LayoutParams();
			layoutParams.copyFrom(progressDialog.getWindow().getAttributes());
			layoutParams.width = LinearLayout.LayoutParams.WRAP_CONTENT;
			layoutParams.height = LinearLayout.LayoutParams.WRAP_CONTENT;
			progressDialog.getWindow().setAttributes(layoutParams);
		}
	}

	private void dismissProgressDialog () { if (progressDialog.isShowing()) progressDialog.dismiss(); }

	@Override
	protected void onActivityResult (int requestCode, int resultCode, Intent data) {
		super.onActivityResult(requestCode, resultCode, data);
		mCallbackManager.onActivityResult(requestCode, resultCode, data);
	}

	@Override
	protected void onStart () {
		super.onStart();
		mFirebaseAuth.addAuthStateListener(mAuthStateListener);
	}

	@Override
	protected void onStop () {
		super.onStop();
		if (mAuthStateListener != null) mFirebaseAuth.removeAuthStateListener(mAuthStateListener);
	}

	@Override
	protected void onDestroy () {
		super.onDestroy();
		if (gps!=null) gps.stopUsingGPS();
	}

	//adding user to firestore in other thread
//	private static class AddUserToFirestore extends AsyncTask<Void, Void, Void> {
//		private WeakReference<AuthActivity> activityReference;
//
//		// only retain a weak reference to the activity
//		AddUserToFirestore(AuthActivity context) {
//			activityReference = new WeakReference<>(context);
//		}
//
//		@Override
//		protected void onPreExecute () {
//			super.onPreExecute();
//			activityReference.get().pb_button.startAnim();
//		}
//
//		@Override
//		protected Void doInBackground(Void... params) {
//			activityReference.get().addUserToFirestore();
//			return null;
//		}
//
//		@Override
//		protected void onPostExecute (Void aVoid) {
//			super.onPostExecute(aVoid);
//			activityReference.get().pb_button.stopAnim(() -> activityReference.get().startMainActivity());
//		}
//	}

}

File: app/src/main/java/com/mmdev/meetapp/ui/auth/AuthActivity.kt
package com.mmdev.meetapp.ui.auth

import android.app.Dialog
import android.content.Intent
import android.graphics.Color
import android.os.Bundle
import android.view.Gravity
import android.view.WindowManager
import android.widget.*
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.ViewModelProviders
import co.ceryle.segmentedbutton.SegmentedButtonGroup
import com.facebook.AccessToken
import com.facebook.CallbackManager
import com.facebook.FacebookCallback
import com.facebook.FacebookException
import com.facebook.login.LoginResult
import com.facebook.login.widget.LoginButton
import com.google.android.gms.tasks.Task
import com.google.firebase.auth.AuthResult
import com.google.firebase.auth.FacebookAuthProvider
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.firestore.FirebaseFirestore
import com.mmdev.meetapp.R
import com.mmdev.meetapp.models.ProfileModel
import com.mmdev.meetapp.services.GPSTracker
import com.mmdev.meetapp.ui.MainActivity
import com.mmdev.meetapp.ui.main.ProfileViewModel
import com.mmdev.meetapp.utils.uiUtils
import com.mmdev.progressbuttonlib.ProgressButton
import java.util.*


/**
 *
 */
class AuthActivity: AppCompatActivity() {

	private val TAG = "logi"


	private var facebookLogInButton: LoginButton? = null
	private var facebookLoginButtonDelegate: Button? = null

	private var mFirebaseAuth: FirebaseAuth? = null
	private var mAuthStateListener: FirebaseAuth.AuthStateListener? = null
	private var mFirestore: FirebaseFirestore? = null

	private var mFirebaseUser: FirebaseUser? = null
	//Progress dialog for any authentication action
	private var progressDialog: AlertDialog? = null

	private var mCallbackManager: CallbackManager? = null

	private val gps = GPSTracker()
	private var mGender = "male"
	private var mPreferedGender = "male"
	private var mCity = ""
	private var mProfileModel: ProfileModel? = null

	private var progressButton: ProgressButton? = null
	private var additionalRegDialog: Dialog? = null

	private var profileViewModel: ProfileViewModel? = null

	override fun onCreate(savedInstanceState: Bundle?) {
		super.onCreate(savedInstanceState)
		setContentView(R.layout.activity_auth)
		facebookLogInButton = findViewById(R.id.facebook_login_button)
		facebookLoginButtonDelegate = findViewById(R.id.facebook_login_button_delegate)

		mFirebaseAuth = FirebaseAuth.getInstance()
		mAuthStateListener = FirebaseAuth.AuthStateListener {  }
		mFirestore = FirebaseFirestore.getInstance()
		mCallbackManager = CallbackManager.Factory.create()
		setUpFacebookLoginButton()
		setProgressDialog()
		profileViewModel = ViewModelProviders.of(this).get(ProfileViewModel::class.java)

	}

	private fun setUpFacebookLoginButton() {
		facebookLogInButton!!.registerCallback(mCallbackManager, object: FacebookCallback<LoginResult> {
			override fun onSuccess(loginResult: LoginResult) {
				handleFacebookAccessToken(loginResult.accessToken)
			}

			override fun onCancel() {}

			override fun onError(error: FacebookException) {}
		})
		facebookLoginButtonDelegate!!.setOnClickListener { facebookLogInButton!!.performClick() }
	}

	private fun handleFacebookAccessToken(token: AccessToken) {
		showProcessProgressDialog()
		val credential = FacebookAuthProvider.getCredential(token.token)
		mFirebaseAuth!!.signInWithCredential(credential).addOnCompleteListener(this) { task: Task<AuthResult> ->
			if (task.isSuccessful && mFirebaseAuth!!.currentUser != null) {
				mFirebaseUser = mFirebaseAuth!!.currentUser
				handleUserExistence(mFirebaseUser!!.uid)
			}
			else uiUtils.showSafeToast(applicationContext, "LogIn Aborted")
			dismissProgressDialog()
		}
	}

	private fun handleUserExistence(uId: String) {
		val users = mFirestore!!.collection("users")
		users.document(uId).get().addOnCompleteListener { task ->
			if (task.isSuccessful) {
				val document = task.result
				if (document != null && document.exists()) {
					mProfileModel = document.toObject(ProfileModel::class.java)
					profileViewModel!!.saveProfile(applicationContext, mProfileModel)
					profileViewModel!!.setProfileModel(mProfileModel)
					Toast.makeText(applicationContext, "Exist", Toast.LENGTH_SHORT).show()
					startMainActivity()
				}
				else showRegistrationDialog()
			}
			else Toast.makeText(applicationContext, "Can't connect to server", Toast.LENGTH_SHORT).show()
		}
	}

	private fun showRegistrationDialog() {
		additionalRegDialog = Dialog(this)
		additionalRegDialog!!.setContentView(R.layout.dialog_registration)
		val window = additionalRegDialog!!.window
		additionalRegDialog!!.show()
		window?.setLayout(LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.MATCH_PARENT)

		val sbgGender = additionalRegDialog!!.findViewById<SegmentedButtonGroup>(R.id.dialog_registr_sbg_gender)
		val sbgPrefgender =
			additionalRegDialog!!.findViewById<SegmentedButtonGroup>(R.id.dialog_registr_sbg_preferedgender)

		sbgGender.setOnClickedButtonListener { position -> mGender = if (position == 0) "male" else "female" }

		sbgPrefgender.setOnClickedButtonListener { position ->
			when (position) {
				0 -> mPreferedGender = "male"
				1 -> mPreferedGender = "female"
				2 -> mPreferedGender = "both"
			}
		}
		progressButton = additionalRegDialog!!.findViewById(R.id.diag_reg_btn_done)
		progressButton!!.setOnClickListener {
			//new AddUserToFirestore(this).execute();
			progressButton!!.startAnim()
			addUserToFirestore()
		}
	}

	private fun addUserToFirestore() {
		//city = gps.getCity(this);
		mCity = "Kyiv"
		val urls = ArrayList<String>()
		urls.add(mFirebaseUser!!.photoUrl.toString())
		mProfileModel = ProfileModel(mFirebaseUser!!.displayName!!,
		                             mCity,
		                             mGender,
		                             mPreferedGender,
		                             mFirebaseUser!!.photoUrl.toString(),
		                             urls,
		                             mFirebaseUser!!.uid)
		val profiles = mFirestore!!.collection("users")
		profiles.document(mFirebaseUser!!.uid).set(mProfileModel!!).addOnSuccessListener {
			profileViewModel!!.saveProfile(this, mProfileModel)
			profileViewModel!!.setProfileModel(mProfileModel)
			Toast.makeText(this,
			               "Successfully uploaded data to firebase + stored into sharedPrefs",
			               Toast.LENGTH_LONG).show()
			progressButton!!.stopAnim { startMainActivity() }
		}

	}

	private fun startMainActivity() {
		val mMainActivityIntent = Intent(this@AuthActivity, MainActivity::class.java)
		startActivity(mMainActivityIntent)
		if (additionalRegDialog != null) additionalRegDialog!!.dismiss()
		finish()
	}

	/*
	progress dialog
	 */
	private fun setProgressDialog() {
		val llPadding = 10
		val ll = LinearLayout(this)
		ll.orientation = LinearLayout.HORIZONTAL
		ll.setPadding(llPadding, llPadding, llPadding, llPadding)
		ll.gravity = Gravity.CENTER
		val llParam =
			LinearLayout.LayoutParams(LinearLayout.LayoutParams.WRAP_CONTENT, LinearLayout.LayoutParams.WRAP_CONTENT)
		llParam.gravity = Gravity.CENTER
		ll.layoutParams = llParam

		val progressBar = ProgressBar(this)
		progressBar.isIndeterminate = true
		progressBar.setPadding(llPadding, llPadding, llPadding, llPadding)
		progressBar.layoutParams = llParam

		val tvText = TextView(this)
		tvText.text = getString(R.string.progress_dialog_text)
		tvText.setTextColor(Color.BLACK) //low api
		tvText.textSize = 20f
		tvText.layoutParams = llParam

		ll.addView(progressBar)
		ll.addView(tvText)

		val builder = AlertDialog.Builder(this)
		builder.setCancelable(true)
		builder.setView(ll)

		progressDialog = builder.create()

	}

	private fun showProcessProgressDialog() {
		progressDialog!!.show()
		val window = progressDialog!!.window
		if (window != null) {
			val layoutParams = WindowManager.LayoutParams()
			layoutParams.copyFrom(progressDialog!!.window!!.attributes)
			layoutParams.width = LinearLayout.LayoutParams.WRAP_CONTENT
			layoutParams.height = LinearLayout.LayoutParams.WRAP_CONTENT
			progressDialog!!.window!!.attributes = layoutParams
		}
	}

	private fun dismissProgressDialog() {
		if (progressDialog!!.isShowing) progressDialog!!.dismiss()
	}

	override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
		super.onActivityResult(requestCode, resultCode, data)
		mCallbackManager?.onActivityResult(requestCode, resultCode, data)
	}

	override fun onStart() {
		super.onStart()
		mFirebaseAuth!!.addAuthStateListener(mAuthStateListener!!)
	}

	override fun onStop() {
		super.onStop()
		if (mAuthStateListener != null) mFirebaseAuth!!.removeAuthStateListener(mAuthStateListener!!)
	}

	override fun onDestroy() {
		super.onDestroy()
		gps.stopUsingGPS()
	}


	//adding user to firestore in other thread
	//	private static class AddUserToFirestore extends AsyncTask<Void, Void, Void> {
	//		private WeakReference<AuthActivity> activityReference;
	//
	//		// only retain a weak reference to the activity
	//		AddUserToFirestore(AuthActivity context) {
	//			activityReference = new WeakReference<>(context);
	//		}
	//
	//		@Override
	//		protected void onPreExecute () {
	//			super.onPreExecute();
	//			activityReference.get().progressButton.startAnim();
	//		}
	//
	//		@Override
	//		protected Void doInBackground(Void... params) {
	//			activityReference.get().addUserToFirestore();
	//			return null;
	//		}
	//
	//		@Override
	//		protected void onPostExecute (Void aVoid) {
	//			super.onPostExecute(aVoid);
	//			activityReference.get().progressButton.stopAnim(() -> activityReference.get().startMainActivity());
	//		}
	//	}

}

